### PR TITLE
Feature fitness check and self-test

### PR DESCRIPTION
**Pull request by FlorianZipser**
This pull request includes a mechanism to test Pepper module's whether they fullfill some quality requirements. One of these requirements is the check whether PepperImporter.isImportable(...) is implemented. Another requirement is the self-test, which enables Pepper to run the module under real circumstances in a test environment (with a provided test corpus). For more information see the documentation. 

Tried to skip problems causing by formatting. Only a few files left because of formatting.

### CODE REVIEW

**Comment by thomaskrause** [comment]
I updated the branch to point to the new Salt 3.3.0 release and while everything compiles the DoNothingExporter and DoNothingImporter tests fail when executing the self-tests.

File: pepper-doc/src/main/doxygen/developers/developers.md [hunk 1]
```diff
@@ -25,6 +25,25 @@ pepper-parentModule | Is the parent project for all Pepper modules and contains
 To run Pepper only the projects pepper-lib and pepper-framework are necessary. When there is already an OSGi environment only the project pepper-framework is necessary.
 
 
+Create a runnable Pepper instance with Maven
+===
+
+To create a full functional Pepper instance go to PEPPER_HOME and run
+
+\code
+mvn clean package assembly:single install
```
Comment (thomaskrause) [inline pepper-doc/src/main/doxygen/developers/developers.md]: Isn't "package" and "install" duplicated? From testing "mvn clean install" should work as well.

File: pepper-doc/src/main/doxygen/moduleDevelopers/fitnessFeatureDesc.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+This health feature indicates whether a module is runnable on the Pepper platform. 
+To check this, the function @ref org.corpus_tools.pepper.modules.PepperModule.isReadyToStart() is called.
+When your module fails this check, it means that this method returned false.
+
+### HAS_PASSED_SELFTEST
+
+When the module provides a self test, this health feature indicates whether the self test was passed. A self test is a test setup to check whether the module is runnable within a Pepper instance (similar to a system integration test). Therefore the module needs to provide a corpus as an input and an expected corpus. The output corpus then is compared to the expected corpus.
+When your module fails this check, it means that the self test has not been passed.
+For more information on self tests check section HAS_SELFTEST.
```
Comment (thomaskrause) [inline pepper-doc/src/main/doxygen/moduleDevelopers/fitnessFeatureDesc.md]: HAS_SELFTEST section does not exist

**Comment by thomaskrause** [comment]
I updated the branch to 
* merge the current develop branch
* use the released Salt version 3.3.0 instead of a SNAPSHOT version
* fixed the tests

I would merge it into develop as soon as the documentation about "HAS_SELFTEST" is added or should we wait until all modules are updated?

File: pepper-doc/src/main/doxygen/moduleDevelopers/fitnessFeatureDesc.md (hunk 2 above)
Comment (FlorianZipser) [inline pepper-doc/src/main/doxygen/moduleDevelopers/fitnessFeatureDesc.md]: Yes, it does. See file: fitnessFeatureDesc.md : ### HAS_SELFTEST {#selftest}. Is it somewhere linked incorrectly?

**Comment by FlorianZipser** [comment]
I don't know how travis works, is there a way to configure the workflow? The problem why travis fails is the missing artifact pepper-framework:zip. To create this it is necessary to run:
```
mvn clean package assembly:single install
```

this is not exactly the same as:
```
mvn clean install
```
because of the order is important. First we need to package the project to build an assembly from, in the second step also the assembled zip needs to be installed (into local repo). I assume, this is the missing step in travis and why the artifact pepper-framework:zip was not found.
The necessity of this order is caused in the mechanism how Pepper modules provides and accesses resources. Normally a Pepper module comes as a zip file containing a .jar file and a folder having the same name as the jar. This folder includes resources the module needs when running. Pepper ensures the localization of that folder automatically, so that each module can access its resource folder. But in Pepper itself this folder was not part of the assembly. This issue never came up, because Pepper and the core modules which come with Pepper did not need any resources so far. But this has changed with the self-test mechanism. Now the core modules also need a resource folder (which includes the self-test data). Therefore Pepper now needs to include this folder in its assembly. 
So the question is could travis be configured to run the command listed above. 
A next question which comes into my mind is, when travis has problems with it, will maven cenral have the same problem? If yes may be the maven default goal needs to be configured to run the command above.
Any ideas?

**Comment by FlorianZipser** [comment]
Running the fitness check locally with all available modules produces the following output:
``` 
DoNothingImporter............................................................FIT
SaltXMLImporter..............................................................FIT
TextImporter.................................................................FIT
CoraXMLImporter..........................................................HEALTHY
MMAX2Importer............................................................HEALTHY
CoNLLImporter............................................................HEALTHY
ToolboxImporter..........................................................HEALTHY
ElanImporter.............................................................HEALTHY
EXMARaLDAImporter........................................................HEALTHY
Tiger2Importer...........................................................HEALTHY
GateImporter.............................................................HEALTHY
UAMImporter..............................................................HEALTHY
TEIImporter..............................................................HEALTHY
AldtImporter.............................................................HEALTHY
TreetaggerImporter.......................................................HEALTHY
GrAFImporter.............................................................HEALTHY
WolofImporter............................................................HEALTHY
TCFImporter..............................................................HEALTHY
RSTImporter..............................................................HEALTHY
GenericXMLImporter.......................................................HEALTHY
PTBImporter..............................................................HEALTHY
SpreadsheetImporter......................................................HEALTHY
PAULAImporter............................................................HEALTHY
DOTManipulator...............................................................FIT
DoNothingManipulator.........................................................FIT
SaltValidator................................................................FIT
FALKOManipulator.........................................................HEALTHY
Merger..................................................................CRITICAL
TueBaDZManipulator.......................................................HEALTHY
OrderRelationAdder.......................................................HEALTHY
Timeline2Token...........................................................HEALTHY
WolofGlosser.............................................................HEALTHY
Sentencer................................................................HEALTHY
Tokenizer................................................................HEALTHY
DOTExporter..................................................................FIT
DoNothingExporter............................................................FIT
SaltXMLExporter..............................................................FIT
TextExporter.................................................................FIT
MMAX2Exporter............................................................HEALTHY
ANNISExporter...........................................................CRITICAL
CoNLLExporter............................................................HEALTHY
GraphAnnoExporter........................................................HEALTHY
EXMARaLDAExporter........................................................HEALTHY
SaltInfoExporter.........................................................HEALTHY
TreetaggerExporter.......................................................HEALTHY
TCFExporter..............................................................HEALTHY
PTBExporter..............................................................HEALTHY
PAULAExporter............................................................HEALTHY
```

For the modules PAULAImporter, PAULAExporter, EXMARaLDAImporter and EXMARaLDAExporter self-test are provided with the feature branches feat_self-test. Using these versions results in the fitness check result FIT for these modules.
When a module is HEALTHY it would be good enough in the first place, but it is not clear whether it is runnable in Pepper or not. So adapting the modules to provide self-tests can be done step by step. For now an important issue is to adapt the modules ANNISExporter and Merger which produces a CRITICAL. I will check this in the next days (I think I now their reasons).
In conclusion, for me, the result of the fitness check schould be no blocker for the merge (the result is not used so far, but will be in future in jenkins run (hopefully)). When fixing the issue with the failing we can merge.

File: pepper-doc/src/main/doxygen/moduleDevelopers/fitnessFeatureDesc.md (hunk 2 above)
Comment (thomaskrause) [inline pepper-doc/src/main/doxygen/moduleDevelopers/fitnessFeatureDesc.md]: Its in the markdown page but this section is missing from the generated HTML, at least with doxygen 1.8.11.

**Comment by thomaskrause** [comment]
You can configure Travis to run any maven command, but I would strongly advise against this.

1.  everyone knows that Maven will execute everything when "mvn install" is called, but no one will read the documentation for the correct line
2. Maven Lifecycles don't work that way and we are producing very very ugly Maven builds
3. Everything that makes things more complicated to build will scare developers off (also IDE Maven integration will break)

We should really check if we can integrate everything in  a way that makes "mvn install" possible

**Comment by FlorianZipser** [comment]
Ok, I agree to your point to try things simple. So I applied the building of the assembly to the package phase for the project pepper-framework. This shrinks the maven build from:
```
mvn clean package assembly:single install
```
to 
``` 
mvn clean install assembly:single
```
as it was before. This fixes the travis issue, since it is possible to build the project with 
``` 
mvn install
```
Still, to have a runnable pepper instance you need to do the assembly after building.
